google_guest_agent/command/command.go

```diff
@@ -0,0 +1,140 @@
+type Handler func([]byte) ([]byte, error)
+
+// Request is the basic request structure. Command determines which handler the
+// request is routed to. Callers may set additional arbitrary fields.
+type Request struct {
+	Command string
+}
+
+// Response is the basic response structure. Handlers may set additional
+// arbitrary fields.
+type Response struct {
+	// Status code for the request. Meaning is defined by the caller, but
+	// conventially zero is success.
+	Status int
+	// StatusMessage is an optional message defined by the caller. Should generally
+	// help a human understand what happened.
+	StatusMessage string
+}
+
+var (
+	// CmdNotFoundError is return when there is no handler for the request command
+	CmdNotFoundError = Response{
+		Status:        101,
+		StatusMessage: "Could not find a handler for the requested command",
+	}
+	// BadRequestError is returned for invalid or unparseable JSON
+	BadRequestError = Response{
+		Status:        102,
+		StatusMessage: "Could not parse valid JSON from request",
+	}
+	// ConnError is returned for errors from the underlying communication protocol
+	ConnError = Response{
+		Status:        103,
+		StatusMessage: "Connection error",
+	}
+	// TimeoutError is returned when the timeout period elapses before valid JSON is receieved
+	TimeoutError = Response{
+		Status:        104,
+		StatusMessage: "Connection timeout before reading valid request",
+	}
+	// HandlerError is returned when the handler function returns an non-nil error. The status message will be replaced with the returnd error string.
+	HandlerError = Response{
+		Status:        105,
+		StatusMessage: "The command handler encountered an error processing your request",
+	}
+	// InternalErrorCode is the error code for internal command server errors. Returned when failing to marshal a response.
+	InternalErrorCode = 106
+	internalError     = []byte(`{"Status":106,"StatusMessage":"The command server encountered an internal error trying to respond to your request"}`)
+	handlersMu        = new(sync.RWMutex)
+	handlers          = make(map[string]Handler)
+)
+
+// RegisterHandler registers f as the handler for cmd. If a command.Server has
+// been initialized, it will be signalled to start listening for commands.
+func RegisterHandler(cmd string, f Handler) error {
+	if _, ok := handlers[cmd]; ok {
+		return fmt.Errorf("cmd %s is already handled", cmd)
+	}
+	handlersMu.Lock()
+	defer handlersMu.Unlock()
+	handlers[cmd] = f
+	if cmdserver != nil {
+		cmdserver.Start()
+	}
+	return nil
+}
+
+// UnregisterHandler clears the handlers for cmd. If a command.Server has been
+// intialized and there are no more handlers registered, the server will be
+// signalled to stop listening for commands.
+func UnregisterHandler(cmd string) error {
+	if _, ok := handlers[cmd]; !ok {
+		return fmt.Errorf("cmd %s is not registered", cmd)
+	}
+	handlersMu.Lock()
+	defer handlersMu.Unlock()
+	delete(handlers, cmd)
+	if len(handlers) == 0 && cmdserver != nil {
```

Comment:
I'm not sure we want to stop the server in this condition, we don't have the use case now, but this kind of interface expects that the program may want to register handlers at any time, otherwise a registering code would have to check if the server is running or not. The bottomline is: the main goroutine initializes the server and leaves it functional for any component to register handlers at any point, it will only shut it down when the program is aborted/killed etc.

Reply:
>otherwise a registering code would have to check if the server is running or not

Not really, if we just take out `Start` and `Stop`, replace them with a `CommandServerEnabled` toggle, and listen for the entire server lifetime will still work as expected and allowing registration and deregistration at any time. The commandServer will pick up new handlers on each new connection.